File: server/parser.py
"""
Parse Opus Magnum .solution and .puzzle binary files.

Solution binary format (from omsp Formats.md):
  INT (4 bytes LE): solution format version (= 7)
  STRING: puzzle file name  (length as 7-bit variable-length int, then UTF-8 bytes)
  STRING: solution name
  ...

Puzzle binary format:
  INT (4 bytes LE): puzzle format version (= 3)
  STRING: puzzle name  (human-readable, e.g. "Stabilized Water")
  ...
"""

import struct


def _read_varint(data: bytes, offset: int) -> tuple[int, int]:
    """Read a C# BinaryReader-style 7-bit encoded int. Returns (value, new_offset)."""
    result = 0
    shift = 0
    while True:
        b = data[offset]
        offset += 1
        result |= (b & 0x7F) << shift
        if not (b & 0x80):
            break
        shift += 7
    return result, offset


def _read_string(data: bytes, offset: int) -> tuple[str, int]:
    length, offset = _read_varint(data, offset)
    return data[offset : offset + length].decode("utf-8"), offset + length


def extract_puzzle_id(data: bytes) -> str:
    """Extract the puzzle ID (e.g. 'P008') from a .solution file's bytes."""
    if len(data) < 5:
        raise ValueError("Solution file too short")
    version = struct.unpack_from("<I", data, 0)[0]
    if version != 7:
        raise ValueError(f"Unexpected solution version: {version}")
    puzzle_id, _ = _read_string(data, 4)
    return puzzle_id


def extract_puzzle_name(data: bytes) -> str:
    """Extract the human-readable name (e.g. 'Stabilized Water') from a .puzzle file's bytes."""
    if len(data) < 5:
        raise ValueError("Puzzle file too short")
    name, _ = _read_string(data, 4)
    return name

File: client/src/parser.rs
/// Extract the puzzle ID from a .solution file's bytes.
/// Format: 4-byte LE version (7), then a C#-style 7-bit length-prefixed UTF-8 string.
pub fn extract_puzzle_id(data: &[u8]) -> Option<String> {
    if data.len() < 5 {
        return None;
    }
    let version = u32::from_le_bytes(data[0..4].try_into().ok()?);
    if version != 7 {
        return None;
    }
    let (length, idx) = read_varint(data, 4)?;
    let end = idx + length;
    if end > data.len() {
        return None;
    }
    String::from_utf8(data[idx..end].to_vec()).ok()
}

fn read_varint(data: &[u8], mut offset: usize) -> Option<(usize, usize)> {
    let mut result: usize = 0;
    let mut shift = 0;
    loop {
        if offset >= data.len() {
            return None;
        }
        let b = data[offset] as usize;
        offset += 1;
        result |= (b & 0x7F) << shift;
        if b & 0x80 == 0 {
            break;
        }
        shift += 7;
    }
    Some((result, offset))
}
